RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc

```diff
@@ -0,0 +1,361 @@
+
+    // to cache the TrackInfo
+    std::map<unsigned int, btagbtvdeep::TrackInfoBuilder> trackinfos;
+
+    // unsorted reference to sv
+    const auto & svs_unsorted = *svs;     
+    // fill collection, from DeepTNtuples plus some styling
+    for (unsigned int i = 0; i <  jet.numberOfDaughters(); i++){
+        auto cand = jet.daughter(i);
+        if(cand){
+          // candidates under 950MeV (configurable) are not considered
+          // might change if we use also white-listing
+          if (cand->pt()< min_candidate_pt_) continue; 
+          if (cand->charge() != 0) {
+            auto & trackinfo = trackinfos.emplace(i,track_builder).first->second;
+            trackinfo.buildTrackInfo(cand,jet_dir,jet_ref_track_dir,pv);
+            c_sorted.emplace_back(i, trackinfo.getTrackSip2dSig(),
+                    -btagbtvdeep::mindrsvpfcand(svs_unsorted,cand), cand->pt()/jet.pt());
+          } else {
+            n_sorted.emplace_back(i, -1,
+                    -btagbtvdeep::mindrsvpfcand(svs_unsorted,cand), cand->pt()/jet.pt());
+          }
+        }
+    }
+
+    // sort collections (open the black-box if you please) 
+    std::sort(c_sorted.begin(),c_sorted.end(),
+      btagbtvdeep::SortingClass<std::size_t>::compareByABCInv);
+    std::sort(n_sorted.begin(),n_sorted.end(),
+      btagbtvdeep::SortingClass<std::size_t>::compareByABCInv);
+
+    std::vector<size_t> c_sortedindices,n_sortedindices;
+   
+    // this puts 0 everywhere and the right position in ind 
+    c_sortedindices=btagbtvdeep::invertSortingVector(c_sorted);
+    n_sortedindices=btagbtvdeep::invertSortingVector(n_sorted);
+
+    // set right size to vectors
+    features.c_pf_features.clear();
+    features.c_pf_features.resize(c_sorted.size());
+    features.n_pf_features.clear();
+    features.n_pf_features.resize(n_sorted.size());
+
+
+
+  for (unsigned int i = 0; i <  jet.numberOfDaughters(); i++){
+
+    // get pointer and check that is correct
+    auto cand = dynamic_cast<const reco::Candidate *>(jet.daughter(i));
+    if(!cand) continue;
+    // candidates under 950MeV are not considered
+    // might change if we use also white-listing
+    if (cand->pt()<0.95) continue;
+
+    auto packed_cand = dynamic_cast<const pat::PackedCandidate *>(cand);
+    auto reco_cand = dynamic_cast<const reco::PFCandidate *>(cand);
+
+    // need some edm::Ptr or edm::Ref if reco candidates
+    reco::PFCandidatePtr reco_ptr;
+    if (pf_jet) {
+      reco_ptr = pf_jet->getPFConstituent(i);
+    } else if (pat_jet && reco_cand) {
+      reco_ptr = pat_jet->getPFConstituent(i);
+    }
+    // get PUPPI weight from value map
+    float puppiw = 1.0; // fallback value
+    if (reco_cand && use_puppi_value_map_) {
+      puppiw = (*puppi_value_map)[reco_ptr];
+    } else if (reco_cand) {
+      edm::LogWarning("MissingFeatures") << "PUPPI value map missing. "
+                                         << puppiw << " will be used as default";
```

Comment:
please limit this to one printout per module instance to avoid flooding the logs.
OTOH, it may be better to reduce heuristics and add a configuration, e.g. fallBackPuppiWeightTo1, and then if it's false and we are here, make an exception.